pkg/datapath/linux/config/config.go

```diff
@@ -449,6 +449,11 @@ func (h *HeaderfileWriter) writeTemplateConfig(fw *bufio.Writer, e datapath.Endp
 		ctmap.WriteBPFMacros(fw, nil)
 	}
 
+	// The IPv4 gateway of the endpoint.
+	// While IPV4_GATEWAY can change when cilium agent reloads, LXC_IPV4_GATEWAY
+	// is fixed with the value at the moment the endpoint was created.
```

Comment:
Probably not my best comment here. So we track the gateway per endpoint because the endpoints are created at different times. But the gateway does not ever change for a given waypoint. And also we don't expect many different gateways around globally, as for it to change is usually avoided to the best of cilium's ability. So we don't end up with a template per endpoint, just a template per gateway.

I think it works in both places and it's a question of whether we prefer having as many templates as different gateways vs substituting on each endpoint restore. Agree that it looks more reasonable as static data if we did not care about any of the options, but thought that it would be preferable to have as many templates as different gateways.

Reply:
> So we don't end up with a template per endpoint, just a template per gateway.

:+1: Good point, it may only change for every restart (possibly less if we carry the gateway across restart, but I don't think we make any such guarantee today).

>  if we did not care about any of the options

Which other options?

> thought that it would be preferable to have as many templates as different gateways.

This part I don't understand. What advantage do you see for having more templates? The whole idea with templates is to reduce the number of required compiles to ideally 1 per cilium run. If endpoints end up running for a long time, worst case it becomes 1 per cilium restart.

Reply:
I don't have a good reason really. I just thought that unless it is strictly a property that changes per endpoint, it goes on the template. I saw it as if, replacing symbols in the elf is something hacky that is done because it is needed but avoided if possible. Anyway, not good reason, indeed is less templates if we add it to the static data, so good there.

Reply:
Hacky is in the eye of the beholder :-) 

We rely heavily on replacing symbols in the ELF. Adding one more doesn't move the needle much, we still iterate through all data in the ELF, rewriting to another location and substituting variables as we do so.  Sure there may be less hacky ways to achieve static data substitution, but those are longer-term objectives and even if we improve the underlying implementation, we should *still* use some form of static data substitution to minimize recompilation requirements. That won't change even if we add more static data values. For now, I'm happy to move as much into static data as we can, particularly if it really is static data - ie just values that may change between endpoints, but won't change at runtime.

FYI, using static data for conditionals in the datapath is another separate case, but the tradeoff is different there because it can increase verification complexity since the eBPF code could take different execution paths depending on the configured value. Not directly relevant here on this PR but for those I think we'd need to look closer on a case-by-case basis.